integration-tests/tests/src/tests/sync/sync-session.ts

```diff
@@ -0,0 +1,1415 @@
+      const partition = generatePartition();
+      let realm!: Realm;
+      return this.app
+        .logIn(Realm.Credentials.anonymous())
+        .then((user) => {
+          const config = getSyncConfiguration(user, partition);
+          realm = new Realm(config);
+          return realm.syncSession?.downloadAllServerChanges(1);
+        })
+        .then(
+          () => {
+            throw new Error("Download did not time out");
+          },
+          (e) => {
+            expect(e).equals("Downloading changes did not complete in 1 ms.");
+            return realm.syncSession?.downloadAllServerChanges();
+          },
+        );
+    });
+
+    it("timeout on upload successfully throws", async function (this: AppContext) {
+      let realm!: Realm;
+      const partition = generatePartition();
+      return this.app
+        .logIn(Realm.Credentials.anonymous())
+        .then((user) => {
+          const config = getSyncConfiguration(user, partition);
+          realm = new Realm(config);
+          return realm.syncSession?.uploadAllLocalChanges(1);
+        })
+        .then(
+          () => {
+            throw new Error("Upload did not time out");
+          },
+          (e) => {
+            expect(e).equals("Uploading changes did not complete in 1 ms.");
+            return realm.syncSession?.uploadAllLocalChanges();
+          },
+        );
+    });
+  });
+
+  describe("reconnect", () => {
+    afterEach(() => Realm.clearTestState());
+    it("smoketest", async function (this: AppContext) {
+      const credentials = Realm.Credentials.anonymous();
+      const partition = generatePartition();
+      return this.app.logIn(credentials).then((user) => {
+        const config = getSyncConfiguration(user, partition);
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const realm = new Realm(config);
+
+        // No real way to check if this works automatically.
+        // This is just a smoke test, making sure the method doesn't crash outright.
+        Realm.App.Sync.reconnect(this.app);
+      });
+    });
+  });
+
+  describe("hasExistingSessions", () => {
+    afterEach(() => Realm.clearTestState());
+    it("starting and stopping syncsessions propagates to hasExistingSessions", async function (this: AppContext) {
+      expect(Realm.App.Sync._hasExistingSessions(this.app)).to.be.false;
+
+      const credentials = Realm.Credentials.anonymous();
+      const partition = generatePartition();
+      return this.app.logIn(credentials).then((user) => {
+        const config = getSyncConfiguration(user, partition);
+        const realm = new Realm(config);
+        expect(Realm.App.Sync._hasExistingSessions(this.app)).to.be.true;
+        realm.close();
+
+        // Wait for the session to finish
+        return async () => {
+          for (let i = 50; i >= 0; i--) {
+            if (!Realm.App.Sync._hasExistingSessions(this.app)) {
+              return;
+            }
+            sleep(100);
```

Comment:
A promise like this must be awaited.

```suggestion
            await sleep(100);
```